### PR TITLE
refactor(grapher): better handling of query params

### PR DESCRIPTION
**Pull request by marcelgerber**
The way we were handling grapher query params was subpar at best - including that some of the types were not an accurate reflection of the properties that can actually be set, and so on.

https://github.com/owid/owid-grapher/commit/d2da674f1599db3ee1a9016c6ae3df2a74d02006 's commit message explains the changes made to the `GrapherQueryParams` type, for example.
There were also some corners where we had less type-safety than we can have.

`grapherConfigToQueryParams` is not used yet, but I'll need it down the road for narrative views. It takes a (partial) chart config and converts it into the query params fitting to achieve this grapher state.

### CODE REVIEW

**Comment by marcelgerber** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4248?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4248** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4248?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4248?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#4208** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4208?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4201** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: site/search/SearchUtils.tsx [hunk 1]
```diff
@@ -121,12 +121,10 @@ export const getEntityQueryStr = (
...
+        return Url.fromQueryStr(existingQueryStr).updateQueryParams({
             // If we have any entities pre-selected, we want to show the chart tab
-            Url.fromQueryStr(existingQueryStr).updateQueryParams({
-                tab: "chart",
-            }),
-            entities
-        ).queryStr
+            tab: "chart",
+            country: generateSelectedEntityNamesParam(entities),
+        } satisfies GrapherQueryParams).queryStr
```
Comment (sophiamersmann) [inline site/search/SearchUtils.tsx:128]: I had to read up on `satisfies` (again!). Beautiful to see it in action :)

**Review by sophiamersmann — APPROVED** [review]
Very nice! 🙌🏻

File: site/search/SearchUtils.tsx (hunk 1 above)
Comment (marcelgerber) [reply site/search/SearchUtils.tsx:128]: It's also my first use of it 🎉 

In this case, it's just a way to ensure that the object actually matches that type, and to do that inline. The alternative would be to use `as`, but that just blindly casts it to that type and only checks the rough shape, not that it actually only specifies known properties.

So, yeah, a nice and useful addition to TS :)

**Pull request merged**